### Data Cleaning | Guided Project | Exploring Ebay Car Sales Data

Using a used car dataset provided by Dataset (originally scraped and shared on Kaggle by orgesleka), this project aims to clean the data and analyze the car listings.

The data dictionary provided by Dataquest is as follows: 

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [237]:
import pandas as pd
import numpy as np

In [238]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")                 

In [239]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

Let's get an overview of the dataset by using the DataFrame.info() and DataFrame.head() functions

In [240]:
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

From the result, we can observe that: 
* some columns such as "gearbox", "model", "FuelType" contain null values
* the data uses camel case instead of Python's preferred snake case 
* data type for numeric columns such as "price", "yearOfRegistration", "kilometer", "monthOfRegistration"is an integer
* name of the cars contains mispellings and special characters
* types of gearbox are indicated in German ('manuell' for 'manual' and 'automatik' for 'automatic')
* likewise for column "notRepairedDamange" ('ja' for 'yes' and 'nein' for 'no')

### Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [241]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [242]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 
                 'registration_year', 'gearbox', 'power_PS', 'model', 'kilometer', 'registration_month', 
                 'fuel_type', 'brand,', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 
                 'last_seen']

In [243]:
print(autos.head())

          date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_PS  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type      brand, unrepai

### Data Exploration: Let's explore what information is useful and look for numeric data that may be stored as strings or texts that can be converted and cleaned

In [244]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_PS  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type      brand,  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

In [245]:
autos['seller'].unique()

array(['privat', 'gewerblich'], dtype=object)

In [246]:
autos['fuel_type'].unique()

array(['benzin', 'diesel', nan, 'lpg', 'andere', 'hybrid', 'cng',
       'elektro'], dtype=object)

In [247]:
autos['offer_type'].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [248]:
autos['price'].value_counts(5)

0         0.029010
500       0.015261
1500      0.014518
1000      0.012513
1200      0.012365
            ...   
23456     0.000003
171000    0.000003
21830     0.000003
13485     0.000003
8188      0.000003
Name: price, Length: 5597, dtype: float64

In [249]:
autos['nr_of_pictures'].head()

0    0
1    0
2    0
3    0
4    0
Name: nr_of_pictures, dtype: int64

In [250]:
autos['nr_of_pictures'].value_counts()

0    371528
Name: nr_of_pictures, dtype: int64

* Column 'nr_of_pictures' looks odd
* For columns 'seller' and 'offer type', near all dataset are uniform

### Converting text to numeric value

As previously mentioned, since 'price' and 'kilometer' are numeric values but stored as text, we will convert the data to float/integer

In [251]:
autos['price'].astype(float)

0           480.0
1         18300.0
2          9800.0
3          1500.0
4          3600.0
           ...   
371523     2200.0
371524     1199.0
371525     9200.0
371526     3400.0
371527    28990.0
Name: price, Length: 371528, dtype: float64

In [252]:
autos['kilometer'].astype(float)

0         150000.0
1         125000.0
2         125000.0
3         150000.0
4          90000.0
            ...   
371523     20000.0
371524    125000.0
371525    150000.0
371526    150000.0
371527     50000.0
Name: kilometer, Length: 371528, dtype: float64

We will rename 'kilometer' to 'odometer_km'

In [253]:
autos.rename({'kilometer':'odometer_km'}, axis=1, inplace=True)

In [254]:
autos['odometer_km']

0         150000
1         125000
2         125000
3         150000
4          90000
           ...  
371523     20000
371524    125000
371525    150000
371526    150000
371527     50000
Name: odometer_km, Length: 371528, dtype: int64

### Further Cleaning: Since It looks like the nbr_of_photos column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [255]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

In [256]:
autos

date_crawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        price   abtest vehicle_type  registration_year    gearbox  power_PS  \
0         480     test          NaN               1993    manuell         0   
1       18300     test        coupe               2011    manuell       190   
2        9800     test          suv               2004  automatik       163   
3        1500     test   kleinwagen               2001    manuell        75   
4        3600     test   kleinwagen               2008    manuell        69   
...       ...      ...          ...                ...        ...       ...   
371523   2200     test          NaN               2005        NaN         0   
371524   1199     test       cabrio               2000  automatik       101   
371525   9200     test          bus               1996    manuell       102   
371526   3400     test        kombi               2002    manuell       100   
371527  28990  control    limousine               2013    manuell       320   

              model  odometer_km  registration_month fuel_type  \
0              golf       150000                   0    benzin   
1               NaN       125000                   5    diesel   
2             grand       125000                   8    diesel   
3              golf       150000                   6    benzin   
4             fabia        90000                   7    diesel   
...             ...          ...                 ...       ...   
371523          NaN        20000                   1       NaN   
371524       fortwo       125000                   3    benzin   
371525  transporter       150000                   3    diesel   
371526         golf       150000                   6    diesel   
371527      m_reihe        50000                   8    benzin   

                brand, unrepaired_damage           ad_created  postal_code  \
0           volkswagen               NaN  2016-03-24 00:00:00        70435   
1                 audi                ja  2016-03-24 00:00:00        66954   
2                 jeep               NaN  2016-03-14 00:00:00        90480   
3           volkswagen              nein  2016-03-17 00:00:00        91074   
4                skoda              nein  2016-03-31 00:00:00        60437   
...                ...               ...                  ...          ...   
371523  sonstige_autos               NaN  2016-03-14 00:00:00        39576   
371524           smart              nein  2016-03-05 00:00:00        26135   
371525      volkswagen              nein  2016-03-19 00:00:00        87439   
371526      volkswagen               NaN  2016-03-20 00:00:00        40764   
371527             bmw              nein  2016-03-07 00:00:00        73326   

                  last_seen  
0       2016-04-07 03:16:57  
1       2016-04-07 01:46:50  
2       2016-04-05 12:47:46  
3       2016-03-17 17:40:17  
4       2016-04-06 10:17:21  
...                     ...  
371523  2016-04-06 00:46:52  
371524  2016-03-11 18:17:12  
371525  2016-04-07 07:15:26  
371526  2016-03-24 12:45:21  
371527  2016-03-22 03:17:10  

[371528 rows x 17 columns]

### Price and Odometer

As we continue to explore the data, we will remove any outliers or strange values from the dataset

In [257]:
autos['odometer_km'].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

In [258]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
autos['price'].value_counts().head(20)

(5597,)
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

There are 10,778 cars listed with 0 price, given that eBay is an auction site, there could legitimately be items where the opening bid is 0. We will keep them but remove anything above 350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [259]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

In [260]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

In [261]:
autos = autos[autos["price"].between(1, 350000)]
autos["price"].describe()

count    360635.000000
mean       5898.671956
std        8866.359669
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Converting Column Type for dates to integer 

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

In [262]:
autos["registration_year"].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

In [263]:
autos["registration_month"].describe()

count    360635.000000
mean          5.796032
std           3.684416
min           0.000000
25%           3.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64

In [264]:
autos[["date_crawled", "last_seen", "ad_created"]][0:5]

date_crawled            last_seen           ad_created
0  2016-03-24 11:52:17  2016-04-07 03:16:57  2016-03-24 00:00:00
1  2016-03-24 10:58:45  2016-04-07 01:46:50  2016-03-24 00:00:00
2  2016-03-14 12:52:21  2016-04-05 12:47:46  2016-03-14 00:00:00
3  2016-03-17 16:54:04  2016-03-17 17:40:17  2016-03-17 00:00:00
4  2016-03-31 17:25:20  2016-04-06 10:17:21  2016-03-31 00:00:00

Since the first ten digits of the column "date_crawled" represent the data we need (date in numberical value), we will extract these by using the str[] function

In [265]:
autos["date_crawled"].str[:10]

0         2016-03-24
1         2016-03-24
2         2016-03-14
3         2016-03-17
4         2016-03-31
             ...    
371523    2016-03-14
371524    2016-03-05
371525    2016-03-19
371526    2016-03-20
371527    2016-03-07
Name: date_crawled, Length: 360635, dtype: object

In [266]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index())

2016-03-05    0.025547
2016-03-06    0.014483
2016-03-07    0.035657
2016-03-08    0.033469
2016-03-09    0.034115
2016-03-10    0.032645
2016-03-11    0.032773
2016-03-12    0.036242
2016-03-13    0.015783
2016-03-14    0.036330
2016-03-15    0.033424
2016-03-16    0.030205
2016-03-17    0.031647
2016-03-18    0.013119
2016-03-19    0.035271
2016-03-20    0.036400
2016-03-21    0.035682
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032800
2016-03-26    0.031974
2016-03-27    0.030227
2016-03-28    0.035063
2016-03-29    0.034126
2016-03-30    0.033535
2016-03-31    0.031872
2016-04-01    0.034145
2016-04-02    0.035094
2016-04-03    0.038812
2016-04-04    0.037628
2016-04-05    0.012780
2016-04-06    0.003128
2016-04-07    0.001617
Name: date_crawled, dtype: float64

In [267]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values())

2016-04-07    0.001617
2016-04-06    0.003128
2016-04-05    0.012780
2016-03-18    0.013119
2016-03-06    0.014483
2016-03-13    0.015783
2016-03-05    0.025547
2016-03-24    0.029914
2016-03-16    0.030205
2016-03-27    0.030227
2016-03-17    0.031647
2016-03-31    0.031872
2016-03-26    0.031974
2016-03-23    0.032002
2016-03-22    0.032493
2016-03-10    0.032645
2016-03-11    0.032773
2016-03-25    0.032800
2016-03-15    0.033424
2016-03-08    0.033469
2016-03-30    0.033535
2016-03-09    0.034115
2016-03-29    0.034126
2016-04-01    0.034145
2016-03-28    0.035063
2016-04-02    0.035094
2016-03-19    0.035271
2016-03-07    0.035657
2016-03-21    0.035682
2016-03-12    0.036242
2016-03-14    0.036330
2016-03-20    0.036400
2016-04-04    0.037628
2016-04-03    0.038812
Name: date_crawled, dtype: float64

Date crawled mostly ranged within March and April of 2016, more or less uniformly spreadout during this period. 

In [268]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index())

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007939
2016-03-09    0.009824
2016-03-10    0.011460
2016-03-11    0.012955
2016-03-12    0.023240
2016-03-13    0.008410
2016-03-14    0.012176
2016-03-15    0.016324
2016-03-16    0.016418
2016-03-17    0.028699
2016-03-18    0.006888
2016-03-19    0.016330
2016-03-20    0.019884
2016-03-21    0.020026
2016-03-22    0.020508
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.019000
2016-03-26    0.015958
2016-03-27    0.016721
2016-03-28    0.022189
2016-03-29    0.023284
2016-03-30    0.023725
2016-03-31    0.024243
2016-04-01    0.023897
2016-04-02    0.024967
2016-04-03    0.025308
2016-04-04    0.025536
2016-04-05    0.126962
2016-04-06    0.218950
2016-04-07    0.130437
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [269]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(114,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039001
2016-04-04    0.037736
2016-04-05    0.011613
2016-04-06    0.003119
2016-04-07    0.001553
Name: ad_created, Length: 114, dtype: float64

Ad created mostly fall within 1-2 months from the listing date, but some are as old as 2 years. 

### Fixing Incorrect Data in Registration Year

In [270]:
autos["registration_year"].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

From the minimum value 1000 and max value of 9999, we can  tell that there are some incorrect data in "registration_year"

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [271]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [272]:
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.066699
1999    0.063552
2005    0.062669
2006    0.057708
2001    0.056955
2003    0.056557
2004    0.056173
2002    0.054290
2007    0.050499
1998    0.049691
Name: registration_year, dtype: float64

In [273]:
autos["registration_year"].describe()

count    346660.000000
mean       2002.896489
std           7.244719
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: registration_year, dtype: float64

### Exploring Brand 

In [274]:
autos["name"].unique()

array(['Golf_3_1.6', 'A5_Sportback_2.7_Tdi',
       'Jeep_Grand_Cherokee_"Overland"', ...,
       'Smart_smart_leistungssteigerung_100ps', 'VW_Golf_Kombi_1_9l_TDI',
       'BMW_M135i_vollausgestattet_NP_52.720____Euro'], dtype=object)

In [275]:
autos["name"].value_counts(normalize=True)

BMW_318i                                                             0.001809
Ford_Fiesta                                                          0.001788
Volkswagen_Golf_1.4                                                  0.001739
Opel_Corsa                                                           0.001728
BMW_316i                                                             0.001509
                                                                       ...   
Golf_4_Diesel_tausch_gegen_Golf_2/3_gti_vr6                          0.000003
Golf_5_GTi_ca.250PS_Ansaugsystem_19"_Alu_Gewindefahrwerk_PDC_uvm.    0.000003
Peugeot_406_HDi_Platin_so_sieht_ein_gepflegtes_KFZ_aus               0.000003
PASSAT_2.0_TDI_HIGHLINE_CHROM_NAVI_140_ps_Bj2006                     0.000003
Audi_TT_RS_plus_Sport_AGA_20_Zoll_Garantie                           0.000003
Name: name, Length: 214012, dtype: float64

There are lots of brands that don't have a significant percentage of listings, so we will focus on the top five brand names in the datalist. 

In [276]:
brand_counts = autos["name"].value_counts(normalize=True)
top5_brands = brand_counts.head(5).index
print(top5_brands)

Index(['BMW_318i', 'Ford_Fiesta', 'Volkswagen_Golf_1.4', 'Opel_Corsa',
       'BMW_316i'],
      dtype='object')


In [277]:
brand_mean_prices = {}

for name in top5_brands:
    brand_only = autos[autos["name"]== name]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[name] = int(mean_price)
    
brand_mean_prices

{'BMW_318i': 3837,
 'Ford_Fiesta': 1375,
 'Volkswagen_Golf_1.4': 2115,
 'Opel_Corsa': 1382,
 'BMW_316i': 1972}

Of the top 5 brands, BMWs and Volkwagen are distinctively pricier than the other two. 

### Mileage and Price 

In [278]:
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

BMW_318i               3837
Ford_Fiesta            1375
Volkswagen_Golf_1.4    2115
Opel_Corsa             1382
BMW_316i               1972
dtype: int64


In [279]:
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
BMW_318i                   3837
Ford_Fiesta                1375
Volkswagen_Golf_1.4        2115
Opel_Corsa                 1382
BMW_316i                   1972

In [280]:
brand_mean_mileage = {}

for name in top5_brands:
    brands_only = autos[autos["name"]== name]
    mean_mileage = brands_only["odometer_km"].mean()
    brand_mean_mileage[name] = int(mean_mileage)
    
brand_mean_mileage

{'BMW_318i': 137607,
 'Ford_Fiesta': 122000,
 'Volkswagen_Golf_1.4': 138623,
 'Opel_Corsa': 130300,
 'BMW_316i': 145009}

In [281]:
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [282]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
BMW_316i                   145009
Volkswagen_Golf_1.4        138623
BMW_318i                   137607
Opel_Corsa                 130300
Ford_Fiesta                122000

In [284]:
brand_info["mean_price"] = mean_prices
brand_info

mean_prices  mean_price
BMW_316i                  145009        1972
Volkswagen_Golf_1.4       138623        2115
BMW_318i                  137607        3837
Opel_Corsa                130300        1382
Ford_Fiesta               122000        1375

The range of car mileages does not vary as much as the prices do by brand. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.